sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/SessionNotAvailableRetryTest.java

```diff
@@ -0,0 +1,338 @@
+            .directMode()
+            .buildAsyncClient();
+        AsyncDocumentClient asyncDocumentClient = (AsyncDocumentClient) FieldUtils.readField(client,
+            "asyncDocumentClient", true);
+        RxDocumentClientImpl rxDocumentClient = (RxDocumentClientImpl) asyncDocumentClient;
+        GlobalEndpointManager globalEndpointManager =
+            ReflectionUtils.getGlobalEndpointManager(rxDocumentClient);
+        DatabaseAccount databaseAccount = globalEndpointManager.getLatestDatabaseAccount();
+        this.databaseAccount = databaseAccount;
+
+    }
+
+    @DataProvider(name = "preferredRegions")
+    private Object[][] preferredRegions() {
+        List<String> preferredLocations1 = new ArrayList<>();
+        List<String> regionalSuffix1 = new ArrayList<>();
+        List<String> preferredLocations2 = new ArrayList<>();
+        List<String> regionalSuffix2 = new ArrayList<>();
+        Iterator<DatabaseAccountLocation> locationIterator = this.databaseAccount.getReadableLocations().iterator();
+        while (locationIterator.hasNext()) {
+            DatabaseAccountLocation accountLocation = locationIterator.next();
+            preferredLocations1.add(accountLocation.getName());
+            regionalSuffix1.add(getRegionalSuffix(accountLocation.getEndpoint(), TestConfigurations.HOST));
+        }
+
+        //putting preference in opposite direction than what came from database account api
+        for (int i = preferredLocations1.size() - 1; i >= 0; i--) {
+            preferredLocations2.add(preferredLocations1.get(i));
+            regionalSuffix2.add(regionalSuffix1.get(i));
+        }
+
+        return new Object[][]{
+            new Object[]{preferredLocations1, regionalSuffix1, OperationType.Read},
+            new Object[]{preferredLocations2, regionalSuffix2, OperationType.Read},
+            new Object[]{preferredLocations1, regionalSuffix1, OperationType.Query},
+            new Object[]{preferredLocations2, regionalSuffix2, OperationType.Query},
+            new Object[]{preferredLocations1, regionalSuffix1, OperationType.Create},
+            new Object[]{preferredLocations2, regionalSuffix2, OperationType.Create},
+        };
+    }
+
+    @Test(groups = {"multi-master"}, dataProvider = "preferredRegions")
+    public void sessionNotAvailableRetryMultiMaster(List<String> preferredLocations, List<String> regionalSuffix,
+                                                    OperationType operationType) throws IllegalAccessException {
+        CosmosAsyncClient preferredListClient = new CosmosClientBuilder()
+            .endpoint(TestConfigurations.HOST)
+            .key(TestConfigurations.MASTER_KEY)
+            .contentResponseOnWriteEnabled(true)
+            .directMode()
+            .preferredRegions(preferredLocations)
+            .buildAsyncClient();
+
+        AsyncDocumentClient asyncDocumentClient = (AsyncDocumentClient) FieldUtils.readField(preferredListClient,
```

Comment:
we have a helper method, please use this:
ReflectionUtils.getAsyncClient(preferredListClient)

Reply:
done